File: instrumentation/src/test/java/io/opentelemetry/android/internal/features/persistence/DiskManagerTest.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.internal.features.persistence;

import static org.mockito.Mockito.clearInvocations;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.verifyNoMoreInteractions;

import io.opentelemetry.android.config.DiskBufferingConfiguration;
import io.opentelemetry.android.config.OtelRumConfig;
import io.opentelemetry.android.internal.services.CacheStorageService;
import io.opentelemetry.android.internal.services.PreferencesService;
import io.opentelemetry.android.internal.services.ServiceManager;
import java.io.File;
import java.io.IOException;
import java.util.Objects;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.io.TempDir;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class DiskManagerTest {

    private static final String MAX_FOLDER_SIZE_KEY = "max_signal_folder_size";
    @Mock CacheStorageService cacheStorageService;
    @Mock PreferencesService preferencesService;
    @Mock DiskBufferingConfiguration diskBufferingConfiguration;
    @TempDir File cacheDir;
    private DiskManager diskManager;

    @BeforeEach
    public void setUp() {
        OtelRumConfig config = mock(OtelRumConfig.class);
        ServiceManager serviceManager = mock(ServiceManager.class);
        doReturn(diskBufferingConfiguration).when(config).getDiskBufferingConfiguration();
        doReturn(cacheStorageService).when(serviceManager).getService(CacheStorageService.class);
        doReturn(preferencesService).when(serviceManager).getService(PreferencesService.class);
        ServiceManager.setForTest(serviceManager);
        diskManager = DiskManager.create(config);
    }

    @Test
    public void provideSignalBufferDir() throws IOException {
        doReturn(cacheDir).when(cacheStorageService).getCacheDir();
        File expected = new File(cacheDir, "opentelemetry/signals");

        Assertions.assertEquals(expected, diskManager.getSignalsBufferDir());
        Assertions.assertTrue(expected.exists());
    }

    @Test
    public void provideTemporaryDir() throws IOException {
        doReturn(cacheDir).when(cacheStorageService).getCacheDir();
        File expected = new File(cacheDir, "opentelemetry/temp");

        Assertions.assertEquals(expected, diskManager.getTemporaryDir());
        Assertions.assertTrue(expected.exists());
    }

    @Test
    public void cleanupTemporaryDirBeforeProvidingIt() throws IOException {
        File dir = new File(cacheDir, "opentelemetry/temp");
        Assertions.assertTrue(dir.mkdirs());
        Assertions.assertTrue(new File(dir, "somefile.tmp").createNewFile());
        Assertions.assertTrue(new File(dir, "some_other_file.tmp").createNewFile());
        Assertions.assertTrue(new File(dir, "somedir").mkdirs());
        Assertions.assertTrue(new File(dir, "somedir/some_other_file.tmp").createNewFile());

        File temporaryDir = diskManager.getTemporaryDir();

        Assertions.assertTrue(temporaryDir.exists());
        Assertions.assertEquals(0, Objects.requireNonNull(temporaryDir.listFiles()).length);
    }

    @Test
    public void getMaxCacheFileSize() {
        int persistenceSize = 1024 * 1024 * 2;
        doReturn(persistenceSize).when(diskBufferingConfiguration).getMaxCacheFileSize();

        Assertions.assertEquals(persistenceSize, diskManager.getMaxCacheFileSize());

        verify(diskBufferingConfiguration).getMaxCacheFileSize();
    }

    @Test
    public void getMaxSignalFolderSize() {
        long maxCacheSize = 10 * 1024 * 1024; // 10 MB
        int maxCacheFileSize = 1024 * 1024; // 1 MB
        doReturn((int) maxCacheSize).when(diskBufferingConfiguration).getMaxCacheSize();
        doReturn(maxCacheFileSize).when(diskBufferingConfiguration).getMaxCacheFileSize();
        doReturn(maxCacheSize).when(cacheStorageService).ensureCacheSpaceAvailable(maxCacheSize);
        doReturn(-1).when(preferencesService).retrieveInt(MAX_FOLDER_SIZE_KEY, -1);

        // Expects the size of a single signal type folder minus the size of a cache file, to use as
        // temporary space for reading.
        int expected = 2_446_677;
        Assertions.assertEquals(expected, diskManager.getMaxFolderSize());
        verify(preferencesService).store(MAX_FOLDER_SIZE_KEY, expected);

        // On a second call, should get the value from the preferences.
        clearInvocations(cacheStorageService, diskBufferingConfiguration, preferencesService);
        doReturn(expected).when(preferencesService).retrieveInt(MAX_FOLDER_SIZE_KEY, -1);
        Assertions.assertEquals(expected, diskManager.getMaxFolderSize());
        verify(preferencesService).retrieveInt(MAX_FOLDER_SIZE_KEY, -1);
        verifyNoMoreInteractions(preferencesService);
        verifyNoInteractions(cacheStorageService, diskBufferingConfiguration);
    }

    @Test
    public void getMaxSignalFolderSize_whenCalculatedSizeIsNotValid() {
        long maxCacheSize = 1024 * 1024; // 1 MB
        int maxCacheFileSize = 1024 * 1024; // 1 MB
        doReturn((int) maxCacheSize).when(diskBufferingConfiguration).getMaxCacheSize();
        doReturn(maxCacheFileSize).when(diskBufferingConfiguration).getMaxCacheFileSize();
        doReturn(maxCacheSize).when(cacheStorageService).ensureCacheSpaceAvailable(maxCacheSize);
        doReturn(-1).when(preferencesService).retrieveInt(MAX_FOLDER_SIZE_KEY, -1);

        // Expects the size of a single signal type folder minus the size of a cache file, to use as
        // temporary space for reading.
        int expected = 0;
        Assertions.assertEquals(expected, diskManager.getMaxFolderSize());
        verify(preferencesService, never()).store(MAX_FOLDER_SIZE_KEY, expected);
    }
}

File: core/src/test/java/io/opentelemetry/android/internal/features/persistence/DiskManagerTest.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.internal.features.persistence

import io.mockk.Called
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.verify
import io.opentelemetry.android.features.diskbuffering.DiskBufferingConfig
import io.opentelemetry.android.internal.services.CacheStorage
import io.opentelemetry.sdk.testing.assertj.OpenTelemetryAssertions.assertThat
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.File

internal class DiskManagerTest {
    @MockK
    lateinit var cacheStorage: CacheStorage

    @MockK
    lateinit var diskBufferingConfig: DiskBufferingConfig

    @TempDir
    lateinit var cacheDir: File
    private lateinit var diskManager: DiskManager

    @BeforeEach
    fun setUp() {
        MockKAnnotations.init(this)
        every { cacheStorage.cacheDir }.returns(cacheDir)
        diskManager =
            DiskManager(cacheStorage, diskBufferingConfig)
    }

    @Test
    fun `provides the default signal buffer dir if not overridden`() {
        every { diskBufferingConfig.signalsBufferDir } returns null
        val expected = File(cacheDir, "opentelemetry/signals")
        assertThat(diskManager.signalsBufferDir).isEqualTo(expected)
        assertThat(expected.exists()).isTrue()
    }

    @Test
    fun `provides the overridden signal buffer dir`() {
        val customDir = File(cacheDir, "opentelemetry/custom")
        every { diskBufferingConfig.signalsBufferDir } returns customDir
        assertThat(diskManager.signalsBufferDir).isEqualTo(customDir)
        assertThat(customDir.exists()).isTrue()
    }

    @Test
    fun `provides a temp dir`() {
        every { cacheStorage.cacheDir }.returns(cacheDir)
        val expected = File(cacheDir, "opentelemetry/temp")
        assertThat(diskManager.temporaryDir).isEqualTo(expected)
        assertThat(expected.exists()).isTrue()
    }

    @Test
    fun `cleans up the temp dir before providing it`() {
        val dir = File(cacheDir, "opentelemetry/temp")
        assertTrue(dir.mkdirs())
        assertTrue(File(dir, "somefile.tmp").createNewFile())
        assertTrue(File(dir, "some_other_file.tmp").createNewFile())
        assertTrue(File(dir, "somedir").mkdirs())
        assertTrue(File(dir, "somedir/some_other_file.tmp").createNewFile())
        val temporaryDir = diskManager.temporaryDir
        assertThat(temporaryDir.exists()).isTrue()
        assertThat(temporaryDir.listFiles()?.size ?: -1).isZero()
    }

    @Test
    fun `can get the max cache file size`() {
        val persistenceSize = 1024 * 1024 * 2
        every { diskBufferingConfig.maxCacheFileSize }.returns(persistenceSize)
        assertThat(diskManager.maxCacheFileSize).isEqualTo(persistenceSize)
        verify {
            diskBufferingConfig.maxCacheFileSize
        }
    }

    @Test
    fun `can get max signal folder size`() {
        val maxCacheSize = (10 * 1024 * 1024).toLong() // 10 MB
        val maxCacheFileSize = 1024 * 1024 // 1 MB
        every { diskBufferingConfig.maxCacheSize }.returns(maxCacheSize.toInt())
        every { diskBufferingConfig.maxCacheFileSize }.returns(maxCacheFileSize)

        // Expects the size of a single signal type folder, to use as temporary space for reading.
        val expected = (maxCacheSize / 3).toInt()
        assertThat(diskManager.maxFolderSize).isEqualTo(expected)

        verify { cacheStorage wasNot Called }
    }
}
